File: jvm/src/main/scala/com/muchq/scraphics/tracer/Vec3.scala
package com.muchq.scraphics.tracer

case class Vec3(x: Double, y: Double, z: Double):
  def +(o: Vec3): Vec3 =
    Vec3(x + o.x, y + o.y, z + o.z)

  def -(o: Vec3): Vec3 =
    Vec3(x - o.x, y - o.y, z - o.z)

  def length(): Double =
    Math.sqrt(this dot this)

  def *(d: Double): Vec3 =
    Vec3(x * d, y * d, z * d)

  def dot(o: Vec3): Double =
    x * o.x + y * o.y + z * o.z

File: domains/graphics/libs/tracy_go/vec3.go
package tracy

import (
	"math"
)

type Vec3 struct {
	X, Y, Z float64
}

func NewVec3(x, y, z float64) Vec3 {
	return Vec3{X: x, Y: y, Z: z}
}

func (v Vec3) Add(o Vec3) Vec3 {
	return Vec3{X: v.X + o.X, Y: v.Y + o.Y, Z: v.Z + o.Z}
}

func (v Vec3) Sub(o Vec3) Vec3 {
	return Vec3{X: v.X - o.X, Y: v.Y - o.Y, Z: v.Z - o.Z}
}

func (v Vec3) Scale(s float64) Vec3 {
	return Vec3{X: v.X * s, Y: v.Y * s, Z: v.Z * s}
}

func (v Vec3) Dot(o Vec3) float64 {
	return v.X*o.X + v.Y*o.Y + v.Z*o.Z
}

func (v Vec3) Length() float64 {
	return math.Sqrt(v.Dot(v))
}

func (v Vec3) Normalize() Vec3 {
	l := v.Length()
	if l == 0 {
		return v
	}
	return v.Scale(1.0 / l)
}

func (v Vec3) Cross(o Vec3) Vec3 {
	return Vec3{
		X: v.Y*o.Z - v.Z*o.Y,
		Y: v.Z*o.X - v.X*o.Z,
		Z: v.X*o.Y - v.Y*o.X,
	}
}

type Vec2 struct {
	X, Y float64
}

func NewVec2(x, y float64) Vec2 {
	return Vec2{X: x, Y: y}
}